honeybee_grasshopper_energy/src/HB Apply ConstructionSet.py

```diff
@@ -0,0 +1,73 @@
+# Honeybee: A Plugin for Environmental Analysis (GPL)
+# This file is part of Honeybee.
+#
+# Copyright (c) 2019, Ladybug Tools.
+# You should have received a copy of the GNU General Public License
+# along with Honeybee; If not, see <http://www.gnu.org/licenses/>.
+# 
+# @license GPL-3.0+ <http://spdx.org/licenses/GPL-3.0+>
+
+"""
+Apply ConstructionSet to Honeybee Room.
+_
+This component supports the assigning of different construction sets based on
+the Room's average cardinal orientation, provided that a list of ConstructionSets
+are input to the _constr_set. 
+-
+
+    Args:
+        _rooms: Honeybee Rooms to which the input _constr_set should be assigned.
+        _constr_set: A Honeybee ConstructionSet to be applied to the input _room.
+            This can also be text for a construction set to be looked up in the
+            construction set library. If an array of text or construction set
+            objects are input here, different constructions will be assigned
+            based on cardinal direction, starting with north and moving clockwise.
+    
+    Returns:
+        room: The input Rooms with their construction sets edited.
+"""
+
+ghenv.Component.Name = "HB Apply ConstructionSet"
+ghenv.Component.NickName = 'ApplyConstrSet'
+ghenv.Component.Message = '0.1.0'
+ghenv.Component.Category = 'Energy'
+ghenv.Component.SubCategory = '1 :: Constructions'
+ghenv.Component.AdditionalHelpFromDocStrings = '3'
+
+
+try:  # import the honeybee-energy extension
+    from honeybee_energy.lib.constructionsets import construction_set_by_name
+except ImportError as e:
+    raise ImportError('\nFailed to import honeybee_energy:\n\t{}'.format(e))
+
+try:  # import the core honeybee dependencies
+    from honeybee.orientation import angles_from_num_orient, orient_index
+except ImportError as e:
+    raise ImportError('\nFailed to import honeybee:\n\t{}'.format(e))
+
+try:  # import the ladybug_rhino dependencies
+    from ladybug_rhino.grasshopper import all_required_inputs
+except ImportError as e:
+    raise ImportError('\nFailed to import ladybug_rhino:\n\t{}'.format(e))
+
+
+if all_required_inputs(ghenv.Component):
+    # duplicate the initial objects
+    rooms = [obj.duplicate() for obj in _rooms]
+    
+    # process the input constructions
+    for i, constr in enumerate(_constr_set):
+        if isinstance(constr, str):
+            _constr_set[i] = construction_set_by_name(constr)
+    
+    # assign the construction set
+    if len(_constr_set) == 1:  # assign indiscriminately, even if it's a core zone
+        for rm in rooms:
+            rm.properties.energy.construction_set = _constr_set[0]
+    else:  # assign constructions based on cardinal direction
```

Comment:
I think this is a powerful feature but I'm afraid that it will end up being confusing since it is changing the normal expected behavior of Grasshopper component for data matching.

If I connect 10 rooms and 10 construction_set to this component I expect that GH matches each room and construction set together similar to how multiply component would work for two list input. I think we should find a more explicit way to assign them based on cardinal orientations.

I think this is something that we should talk about before I go through the rest of this PR.

Reply:
To be fair, following the "longest combination" data matching scheme that Grasshopper implements whenever you use Item Access instead of List Access is something that we never really supported in any of our legacy components that use List Access because it involves extra code to handle the case that the list lengths don't match.  So I don't think that the argument that Grasshopper logic is sacred does not hold here.

With this said, I know that it's true that we allowed people to input lists of constructions that have the same exact length as the input Zones in legacy. However, in all of the years that I used honeybee-legacy I don't think I ever used that feature once for constructions. And I remember I asked you before I made the "Glazing by Ratio" component for Honeybee[+] whether we should use Grasshopper logic of interpreting a list of glazing ratios as one ratio for each Room/Face and you were pretty clearly on the side of "No" and that, if people need to assign a different glazing ratio for each Room, they can use data trees.

My point is that maybe there are some cases where the Grashopper logic makes sense and maybe there are others where we'd much rather have the option to assign things based on orientation and people can use data trees if they really need full control.  Much like the Glazing ratio, I feel that we want to assign individual constructions based on orientation.  But maybe you have an argument here for ConstructionSets that orientation might not be the main type of organization governing how people usually assign ConstructionSets.